src/app/hooks/useAPI.js

```diff
@@ -0,0 +1,91 @@
+import { useState, useEffect, useRef, useReducer } from 'react';
+import PropTypes from 'prop-types';
+
+import api from '~config/api';
+
+const dataFetchReducer = (state, action) => {
+  switch (action.type) {
+    case 'FETCH_INIT':
+      return {
+        ...state,
+        isLoading: true,
+        isError: ''
+      };
+    case 'FETCH_SUCCESS':
+      return {
+        ...state,
+        response: action.payload,
+        isLoading: false,
+        isError: ''
+      };
+    case 'FETCH_FAILURE':
+      return {
+        ...state,
+        response: action.payload,
+        isLoading: false,
+        isError: 'Something Wrong Happened'
+      };
+    default:
+      return { ...state };
+  }
+};
+
+const useAPI = (config, initialData = {}, execute = false) => {
+  const [state, dispatch] = useReducer(dataFetchReducer, {
+    isLoading: false,
+    isError: '',
+    response: initialData
+  });
+  const [conf, setConf] = useState(config);
+  const exec = useRef(null);
+  exec.current = execute;
+
+  useEffect(() => {
+    let didCancel = false;
+    const fetchData = async () => {
+      dispatch({ type: 'FETCH_INIT' });
+
+      try {
+        const result = await api.any(conf);
+        if (!didCancel) {
+          dispatch({ type: 'FETCH_SUCCESS', payload: result.data });
+        }
+      } catch (error) {
+        if (!didCancel) {
```

Comment:
is necessary make two if  with the same conditional, or maybe you can wrapper all try-catch with only one ?

Reply:
I did it because the idea is that if the request is made, and then the component is removed, no state updates are made. 

If I put it global I would include the request in the conditional and that's not the idea.

what do you think about it?